File: frontend/src/engine/capabilities/namedRange.ts
/**
 * namedRange – Create, update, or delete named ranges.
 *
 * Named ranges can be scoped to the whole workbook (default) or a specific sheet.
 * They are used in formulas (=SUM(SalesData)), validation lists, and dashboards.
 */

import { CapabilityMeta, NamedRangeParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "namedRange",
  description: "Create, update, or delete a named range in the workbook",
  mutates: true,
  affectsFormatting: false,
  // NamedItemCollection.add was introduced in ExcelApi 1.4.
  requiresApiSet: "ExcelApi 1.4",
};

async function handler(
  context: Excel.RequestContext,
  params: NamedRangeParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  const { operation, name, range, sheetName } = params;

  if (options.dryRun) {
    return { stepId: "", status: "success", message: `Would ${operation} named range "${name}"${range ? ` → ${range}` : ""}` };
  }

  if (operation === "delete") {
    options.onProgress?.(`Deleting named range "${name}"...`);
    try {
      const namedItem = context.workbook.names.getItem(name);
      namedItem.delete();
      await context.sync();
    } catch {
      return { stepId: "", status: "success", message: `Named range "${name}" not found — nothing deleted.` };
    }
    return { stepId: "", status: "success", message: `Deleted named range "${name}"` };
  }

  if (!range) {
    return { stepId: "", status: "error", message: "range param is required for create/update." };
  }

  options.onProgress?.(`${operation === "create" ? "Creating" : "Updating"} named range "${name}" → ${range}...`);

  const rng = resolveRange(context, range);
  rng.load("address");
  await context.sync();

  // Build a proper absolute address for the named range formula
  const absAddr = rng.address; // e.g. "'Sheet1'!$A$1:$B$10"

  if (operation === "update") {
    try {
      const existing = context.workbook.names.getItem(name);
      existing.delete();
      await context.sync();
    } catch { /* may not exist yet */ }
  }

  if (sheetName) {
    const ws = context.workbook.worksheets.getItem(sheetName);
    ws.names.add(name, rng);
  } else {
    context.workbook.names.add(name, rng);
  }
  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `${operation === "create" ? "Created" : "Updated"} named range "${name}" → ${absAddr}`,
    outputs: { name, range: absAddr },
  };
}

// ── Legacy-Excel fallback (ExcelApi < 1.4) ────────────────────────────────────
// NamedItemCollection.add requires 1.4. On older builds we degrade to a
// descriptive warning: named ranges are a convenience — any formula that would
// reference the name can use the absolute address directly. We still resolve
// the range so the audit trail captures what the user intended to name.
async function fallback(
  context: Excel.RequestContext,
  params: NamedRangeParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  const { operation, name, range } = params;

  if (operation === "delete") {
    // Deleting a named range that can't exist on this Excel version is a no-op.
    return {
      stepId: "",
      status: "success",
      message:
        `Skipped delete of named range "${name}" — named ranges require ExcelApi 1.4+ ` +
        `so it cannot exist on this Excel version (legacy-Excel fallback).`,
    };
  }

  if (!range) {
    return { stepId: "", status: "error", message: "range param is required for create/update." };
  }

  if (options.dryRun) {
    return {
      stepId: "",
      status: "success",
      message: `Would skip named range "${name}" (not supported on this Excel version).`,
    };
  }

  options.onProgress?.("Legacy-Excel mode: named ranges unavailable, resolving address instead...");

  // Resolve to an absolute address the downstream plan can reference directly.
  const rng = resolveRange(context, range);
  rng.load("address");
  await context.sync();
  const absAddr = rng.address;

  return {
    stepId: "",
    status: "success",
    message:
      `Named range "${name}" not applied — requires ExcelApi 1.4+. Use the ` +
      `absolute address ${absAddr} in formulas that would have referenced the name ` +
      `(legacy-Excel fallback).`,
    outputs: { name, range: absAddr },
  };
}

registry.register(meta, handler as any, fallback as any);
export { meta };

File: backend/app/execution/capabilities/named_range.py
"""namedRange — create / update / delete a workbook or sheet-scoped named range."""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    op = (params.get("operation") or "create").lower()
    name = params.get("name")
    address = params.get("range")
    sheet_scope = params.get("sheetName")
    if not name:
        return {"status": "error", "message": "namedRange requires 'name'."}

    book = ctx.workbook_handle
    if ctx.dry_run:
        return {"status": "preview", "message": f"Would {op} named range {name!r}."}

    try:
        if op == "delete":
            target = book.names[name]
            target.delete()
            return {"status": "success", "message": f"Deleted named range {name!r}.", "outputs": {"name": name}}

        if not address:
            return {"status": "error", "message": "create/update named range requires 'range'."}
        rng = resolve_range(book, address)
        refers_to = f"={rng.sheet.name}!{rng.address}"
        if sheet_scope:
            # Sheet-scoped: prefix name with "Sheet1!Name" — matches Excel's internal format.
            full_name = f"{sheet_scope}!{name}"
        else:
            full_name = name
        # Remove existing with same name before re-creating, for idempotency.
        try:
            book.names[full_name].delete()
        except Exception:
            pass
        book.names.add(full_name, refers_to)
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"{op} failed: {exc}", "error": str(exc)}

    return {
        "status": "success",
        "message": f"{op.capitalize()}d named range {name!r}.",
        "outputs": {"name": name},
    }


registry.register("namedRange", handler, mutates=True)
